### PR TITLE
[GPU] Thread through a common target description

### PR DESCRIPTION
**Pull request by antiagainst**
This commit starts threading through a common target description
for various GPU backends. The goal is to unify how we specify
the target across CUDA, HIP, Vulkan, and others, so that we can
be consistent and reuse the some facilities.

Concretely, we introduce a `#iree_gpu.target` attribute in the
`IREEGPUDialect` to describe the GPU target. It includes a few
fields:

* The canonical target architecture for compilation, e.g.,
  sm_80 for cuda, gfx942 for hip
* A TargetWgpAttr describing the GPU features and limits in a
  single GPU workgroup processsor (that is, AMD compute unit
  or NVIDIA streaming multiprocessor), e.g., compute/storage
  capabilities, subgroup sizes, and workgroup size limits
* An optional TargetChipAttr describing GPU features for the
  final chip or product, e.g., wgp count

To help writing tests, we also introduce another command-line
option, `--iree-gpu-test-target` to specify the target. This option
will be queried if we cannot find the `#iree_gpu.target` attribute
in the wrapping IR. This avoids us to duplicate the full target
for each test to be concise and clear.

In order to support common development targets and the shorthand
attribute, we list some common AMD/NVIDIA GPU architectures
and their capabilities and limits. This is in general fine for
datacenter GPU use cases where we only ever care one architecture.

This starts with update both LLVMGPU backends (CUDA and ROCm)
to use the new target description. Vulkan side is untouched and
the next step to unify.

Progress towards https://github.com/iree-org/iree/issues/16341

ci-extra:  test_nvidia_gpu,test_nvidia_a100,test_amd_mi250,test_amd_w7900

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td [hunk 1]
```diff
@@ -202,5 +312,152 @@ def IREEGPU_MmaScheduleAttr : AttrDef<IREEGPU_Dialect, "MMASchedule"> {
...
+    std::optional<int> getCUDAComputeCapability() const;
+    // Returns true if this target supports TensoreCore MMA ops with TF32
+    // input types.
+    bool supportTF32InputMMAOps() const;
+    // Returns true if this target supports TensorCore synchronized MMA ops.
+    bool supportSyncMMAOps() const;
+  }];
+}
+
+def IREEGPU_AbbrTargetAttr : AttrDef<IREEGPU_Dialect, "AbbrTarget"> {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td]: `Abbr` reads weird  - maybe `TargetAliasAttr` / `TargetNameAttr`?

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td [hunk 2]
```diff
@@ -202,5 +312,152 @@ def IREEGPU_MmaScheduleAttr : AttrDef<IREEGPU_Dialect, "MMASchedule"> {
...
+//===----------------------------------------------------------------------===//
+
+def IREEGPU_TAPI_CUDA    : I32EnumAttrCase<"CUDA",    0, "cuda">;
+def IREEGPU_TAPI_DirectX : I32EnumAttrCase<"DirectX", 1, "directx">;
+def IREEGPU_TAPI_HIP     : I32EnumAttrCase<"HIP",     2, "hip">;
+def IREEGPU_TAPI_Metal   : I32EnumAttrCase<"Metal",   3, "metal">;
+def IREEGPU_TAPI_Vulkan  : I32EnumAttrCase<"Vulkan",  4, "vulkan">;
+def IREEGPU_TAPI_WebGPU  : I32EnumAttrCase<"WebGPU",  5, "webgpu">;
+
+def IREEGPU_TargetAPI : I32EnumAttr<
```
Comment (benvanik) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td]: let's not use enums for things like this - that makes it impossible for out-of-tree/plugins to add targets, unless there's no plan for this to ever exist outside of in-tree code

(and as mentioned I'm sketched out by the fact that this exists - common passes shouldn't be switching on the target API, but instead have whatever API-specific information exists injected via normal MLIR mechanisms like pass options as setup by the target constructing the pipeline)

**Review by benvanik — CHANGES_REQUESTED** [review]
I'm a bit concerned at how rigid this is with the enums - this makes it impossible to reuse the fields in plugins that may add additional information - for example, if a user wanted to experiment out of tree with a different API, a different DotProduct data type, etc they can't without upstreaming the new enum values first. Some of that we could probably fix by turning the enums into proper attribute interfaces (so a user could define their own target API that implemented an interface with whatever you're currently switching on the integer for), but something to keep in mind while designing this stuff: we don't want to end up preventing plugins that interoperate with our lowerings or require lock-step development between the plugins and upstream codebases or nasty patches to inject new enum values.

The enums also have a smell of artificial generality: that you need to know the runtime API is sketchy as it implies there will be common code that will switch on that and make decisions in a way a user cannot control without forking/patching. If common code won't be using the runtime API to switch on then it should be omitted and instead represented by construction: the CUDA target will be constructing all the pipelines and knows it's targeting CUDA, and no common passes it adds to its pipeline should care about the fact that the CUDA target created it and only whatever pass options are specified on it.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td [hunk 3]
```diff
@@ -202,5 +312,152 @@ def IREEGPU_MmaScheduleAttr : AttrDef<IREEGPU_Dialect, "MMASchedule"> {
...
+    "DotProductOpsAttr":$dot,
+    "MMAOpsArrayAttr":$mma,
+
+    // Limits
+    // Supported subgroup size choices.
+    "DenseI32ArrayAttr":$subgroup_size_choices,
+    // The maximal number of threads per X/Y/Z dimension in one workgroup.
+    "DenseI32ArrayAttr":$max_workgroup_sizes,
+    // The maximal number of threads we can have in one workgroup.
+    "uint32_t":$max_thread_size,
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td]: `max_threads_per_workgroup`?

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td [hunk 4]
```diff
@@ -202,5 +312,152 @@ def IREEGPU_MmaScheduleAttr : AttrDef<IREEGPU_Dialect, "MMASchedule"> {
...
+  );
+
+  let assemblyFormat = "`<` struct(params) `>`";
+
+  let extraClassDeclaration = [{
+    int getPreferredSubgroupSize() const {
+      return getCore().getSubgroupSizeChoices().asArrayRef().front();
+    }
+
+    bool supportSubgroupShuffle() const {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td]:
```suggestion
    bool supportsSubgroupShuffle() const {
```

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/test/target_attrs.mlir [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    arch="gfx942",
+    core = <
+      compute = fp16|fp32|int8, storage = b16|b32,
+      subgroup = shuffle|arithmetic, dot = dp4xi8toi32,
+      mma = [<MFMA_F16_16x16x16_F32>, <MFMA_F16_32x32x8_F32>],
+      subgroup_size_choices = [32, 64],
+      max_workgroup_sizes = [1024, 1024, 1024],
+      max_thread_size = 1024,
+      max_workgroup_memory_bytes = 65536>,
+    chip = <core_count = 304>
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/test/target_attrs.mlir]: What do you think about using the RDNA naming: workgroup processor? I think it makes most sense in a device-agnostic attribute; `cores` are ambiguous with cuda cores / tensor cores / matrix cores, etc

**Comment by antiagainst** [comment]
@benvanik: I opted in to use EnumAttr because they makes the C++ code and IR much nicer to work with. I see your points regarding extensibility. I think we can relax to avoid some restrictions brought by it by using normal strings instead. Though stringifying everything doesn't seem a good end result either; so a balance to strike. A good way to go might be adding an additional dictionary to allow plugins inject whatever additional features bits they want. The current ones defined in the TD files are just those we support for ourselves; and hopefully those cover major use cases, given that we have a super broad GPU API and vendor architecture support anyway.

For the target API specifically though, agreed that's overly restrictive. For other enums, I don't expect GPUs to differ that greatly in a sense and our support of various GPU API/vendor/architecture hopefully gets us "comprehensively", this one is entirely big switch.. I think it makes sense to turn it into a string in the abbreviated attribute, and drop it in the full target. It's not that load bearing actually given other bits actually describe the target; it's more informative there.

Does the above sounds reasonable to you? Happy to chat a bit more directly.

**Comment by benvanik** [comment]
Makes sense with the data type enums being ints if you think they will be extended infrequently (like, with new GPU generations) or are fine with adding enum values that out-of-tree users may have even if nothing in-tree uses them. String enums is the way upstream handles things like this even if a bit more annoying to use :( It does allow for entirely out-of-tree work to happen, though, which is nice for private implementations and version shifting (user wants to use a pinned IREE version and even if they land the new enum value upstream would have to then juggle patches/forks/etc). If for a particular enum most are checking it instead of creating it then you can have static isFoo(attr) methods that make the comparison strongly typed (similar to how symbol visibility works with isPublic/etc), or static const char* that can be used for comparisons (`if (stringAttr.getValue() == Foo::Bar)`). Mostly just wanted to call it out because it's a pain to change later on - if you think int is fine then go for it.

Removing the target API seems best. I'm curious why it's needed in any form, though, as it feels duplicative with our target mechanism. Worth a chat so I can understand more about why you want it and if what you want may already be something we can do with those mechanisms or that they can be extended to support for you. For example, TargetBackend's getDefaultExecutableTargets/getHostExecutableTargets takes a deviceID (your target API) as well as its configuration dictionary: https://github.com/iree-org/iree/blob/e4b5a9350aee7cfafefb440a7f01b3d376842d92/compiler/src/iree/compiler/Dialect/HAL/Target/TargetBackend.h#L90-L102
We really don't want to have anything separate from our TargetDevice/TargetBackend as it's very tricky to manage extensibility, multi-targeting (including host/target cross-compilation), etc. Having common utility functions that GPU-based TargetBackend impls can use may be useful, vs. trying to bake it into IR. On my TODO list is a HAL attr that does default target expansion (which our flags will then end up mapping into) and if you're mostly looking to use this for making tests be less verbose that may be something to use.

**Comment by antiagainst** [comment]
> Removing the target API seems best. I'm curious why it's needed in any form, though, as it feels duplicative with our target mechanism.

makes sense. a bit more context why including target api: it makes the target more standalone. but I agree here that we can get the target api from the executable target so better to avoid duplication. So I sent out 60f07cc to drop it entirely. 

(even a bit more background--the target api is somewhat loadbearing particularly for spirv--for example if going down metal/webgpu, we may need different "legalization" patterns/passes for them, e.g., [for webgpu](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SPIRV/Transforms/SPIRVWebGPUTransforms.cpp). not sure we can always put it in a separate pass entry point; sometimes we may need to have "switches" inside existing patterns or so.. a bit sad but software is never perfect i guess..)

> On my TODO list is a HAL attr that does default target expansion (which our flags will then end up mapping into) and if you're mostly looking to use this for making tests be less verbose that may be something to use.

This is why i'm introducing the `iree_gpu.alias_target` in a sense, cause I don't want to duplicate multiple lines of full target description everywhere in the lit tests.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td (hunk 1 above)
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td]: Done.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td (hunk 2 above)
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td]: Dropped now.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/test/target_attrs.mlir (hunk 5 above)
Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/test/target_attrs.mlir]: I basically gave up on finding sanity in GPU terminologies. :D I think i'm fine with whatever. "workgroup processor" is a bit too lengthy though.. "core" is much shorter and good contrast vs. "chip". do you have other good suggestions?

Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/test/target_attrs.mlir]: wg_processor or wgp?

Comment (antiagainst) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/test/target_attrs.mlir]: Changed to use wgp.

**Comment by antiagainst** [comment]
@benvanik I changed to 1) add an extra dictionary attr to allow better flexibility for out of tree targets, and 2) changed to use a cl option `--iree-gpu-test-target` to provide the target for lit tests instead of the alias target, as we chatted before. Could you take another look? thanks.

**Review by benvanik — APPROVED** [review]

**Review by kuhar — APPROVED** [review]
LGTM

**Comment by qedawkins** [comment]
Oh sorry, I think I caused those conflicts because I had the same problem with the enum definitions. Do you want me to fix them?

**Comment by antiagainst** [comment]
> Oh sorry, I think I caused those conflicts because I had the same problem with the enum definitions. Do you want me to fix them?

Don't worry. I think I fixed them. :)

**Comment by qedawkins** [comment]
ok nice :D

**Comment by antiagainst** [comment]
Oh actually new ones.. Let me see. :D

**Pull request merged**